# Simulating the data

Using np.random.choice, generate 100 megabytes(8 bits/byte* 1024 bytes/kilobyte* 1024 kilobytes/megabyte* 100) of random data containing 100%, 90%, 80%, 70%, 60%, and 50% zeros. Be sure to call np.packbits on your data before writing it to a file.   

NC_004681.1:67842-68054 Mycobacterium phage Cjw1, complete genome
NC_011270.1:64264-64767 Mycobacterium phage Spud, complete genome
NC_009817.1:75089-75355 Lactococcus phage KSY1, complete genome


## Generate DNA and protein sequences 100 million letters long

In [26]:
import numpy as np
data_100 = np.random.choice([0, 1], size=8*1024*1024*100, replace=True, p=[1, 0])
data_90 = np.random.choice([0, 1], size=8*1024*1024*100, replace=True, p=[0.9, 0.1])
data_80 = np.random.choice([0, 1], size=8*1024*1024*100, replace=True, p=[0.8, 0.2])
data_70 = np.random.choice([0, 1], size=8*1024*1024*100, replace=True, p=[0.7, 0.3])  
data_60 = np.random.choice([0, 1], size=8*1024*1024*100, replace=True, p=[0.6, 0.4])
data_50 = np.random.choice([0, 1], size=8*1024*1024*100, replace=True, p=[0.5, 0.5])
data_100 = np.packbits(data_100)
data_90 = np.packbits(data_90)
data_80 = np.packbits(data_80)
data_70 = np.packbits(data_70)
data_60 = np.packbits(data_60)
data_50 = np.packbits(data_50)

## Write data to a file

In [28]:
open('zeros_100p', 'wb').write(data_100)
open('zeros_90p', 'wb').write(data_90)
open('zeros_80p', 'wb').write(data_80)
open('zeros_70p', 'wb').write(data_70)
open('zeros_60p', 'wb').write(data_60)
open('zeros_50p', 'wb').write(data_50)

104857600

## Generate DNA and protein sequences 100 million letters long and write those to home directory

In [33]:
DNA_seq = np.random.choice(['A', 'C', 'T', 'G'], size=10000000, replace=True, p=[0.25, 0.25, 0.25, 0.25])
open('nt_seq.fa', 'w').write(''.join(DNA_seq))

10000000

In [35]:
protein_seq = np.random.choice(['A', 'R', 'N', 'D', 'C', 'E', 'Q', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V'], size=10000000, replace=True, p=np.repeat(0.05, 20))
open('protein_seq.fa', 'w').write(''.join(protein_seq))

10000000

## Compressing the data
On each of the files you generated above, run gzip, bzip, pbzip2 and ArithmeticCompress as follows:    
time gzip –k zeros_100p   
time bzip2 –k zeros_100p  
time pbzip2 –k zeros_100p  
time ArithmeticCompresszeros_100p zeros_100p.art

## Table of input file, compression method, size of output file, and time of each command

In [4]:
from IPython.display import HTML, display
import tabulate
table = [["File Name", "Compression method", "File size", "Real time to run (sec)"],
        ["zeros_50p", "-","100 MB", "-"],
         ["zeros_50p.gz", "gzip","101 MB", "3.529s"], 
        ["zeros_50p.bz2", "bzip2","101 MB", "16.895s"],
        ["zeros_50p.bz2", "pbzip2","101 MB", "1.572s"],
        ["zeros_50p.art", "Arithmetic Compress","101 MB", "41.171s"],
        ["zeros_60p", "-","100 MB", "-"],
        ["zeros_60p.gz", "gzip","98 MB", "4.331s"],
        ["zeros_60p.bz2", "bzip2","101 MB", "15.673s"],
        ["zeros_50p.bz2", "pbzip2","101 MB", "1.422s"],
        ["zeros_60p.art", "Arithmetic Compress","98 MB", "41.200s"],
        ["zeros_70p", "-","100 MB", "-"],
        ["zeros_70p.gz", "gzip","90 MB", "6.244s"],
        ["zeros_70p.bz2", "bzip2","96 MB", "14.245s"],
        ["zeros_70p.bz2", "pbzip2","96 MB", "1.180s"],
        ["zeros_70p.art", "Arithmetic Compress","89 MB", "39.299s"],
        ["zeros_80p", "-","100 MB", "-"],
         ["zeros_80p.gz", "gzip","78 MB", "13.352s"], 
        ["zeros_80p.bz2", "bzip2","83 MB", "12.084s"],
        ["zeros_80p.bz2", "pbzip2","83 MB", "0.951s"],
        ["zeros_80p.art", "Arithmetic Compress","73 MB", "35.588s"],
        ["zeros_90p", "-","100 MB", "-"],
         ["zeros_90p.gz", "gzip","57 MB", "18.858s"], 
        ["zeros_90p.bz2", "bzip2","59 MB", "11.466s"],
        ["zeros_90p.bz2", "pbzip2","59 MB", "0.759s"],
        ["zeros_90p.art", "Arithmetic Compress","47 MB", "28.696s"],
        ["zeros_100p", "-","100 MB", "-"],
         ["zeros_100p.gz", "gzip","100 KB", "0.690s"], 
        ["zeros_100p.bz2", "bzip2","8 KB", "0.988s"],
        ["zeros_100p.bz2", "pbzip2","8 KB", "0.096s"],
        ["zeros_100p.art", "Arithmetic Compress","4 KB", "14.768s"]]
display(HTML(tabulate.tabulate(table, tablefmt='html')))

## Questions

_Which algorithm achieves the best level of compression on each file type?_  

    Arithmetic worked best on  bit files, DNA sequence fasta files, and protein sequence fasta files.
       
_Which algorithm is the fastest?_  
    
    pbzip2 was the fastest. 
    
_What is the difference between bzip2 and pbzip2? Do you expect one to be faster and why?_   
    
    pbzip2 uses multi-threading which allows the compression to run at nearly linear speed when the system has multiple CPUs and multi-core processors (slightly more memory to store). Bzip2 relies on the Burrows-Wheeler transform, and then Huffman coding to compress the bytes based on the rarity and occurance of the bytes in the file. bzip2 can only be applied to individual files. 
    pzip2 is expected to run faster since it uses multi-threading. 

_How does the level of compression change as the percentage of zeros increases? Why does this happen?_  

    As percentage of zeroes increase, the level of compression increases. This is due to the decrease in entropy as there are more zeroes. A higher  entropy of a file requires more bits to encode so the level of compression decreases. 
_What is the minimum number of bits required to store a single DNA base?_  
    
    log base 2 (4) = 2 bits
_What is the minimum number of bits required to store an amino acid letter?_    

    log base 2 (20) = 4.32 bits -> 5 bits
_In your tests, how many bits did gzip and bzip2 actually require to store your random DNA and protein sequences?_  
    
    gzip DNA sequence:  
        DNA sequence file: 2.8MB = 2.24 x 10^7 bits  
    bzip2 DNA sequence:      
        DNA sequence file: 2.7MB = 2.16 x 10^7 bits   
    gzip protein sequence:  
        protein sequence file: 5.8MB = 4.64 x 10^7 bits   
    bzip2 protein sequence:  
        protein sequence file: 5.3MB = 4.24 x 10^7 bits   
        
_Are gzip and bzip2 performing well on DNA and proteins?_  

    They performed decently. They both similarly to pbzip2. They performed worse in DNA relative to ArithmeticCompress and better in protein relative to ArithmeticCompress. 
    

Gzip actually rerquired 1 byte (8 bits) to store my random DNA and protein sequences because they read each byte as a char which is worth 1 byte. 

Use real time only because account for multi-core

# DNA

# Protein

In [14]:
table = [["File Name", "Compression method", "File size", "Real time to run (sec)"],
        ["nt_seq.fa", "-","9.6 MB", "-"],
         ["nt_seq.fa.gz", "gzip","2.8 MB", "1.356s"], 
        ["nt_seq.fa.bz2", "bzip2","2.7 MB", "0.952s"],
        ["nt_seq.fa.bz2", "pbzip2","2.7 MB", "0.120s"],
        ["nt_seq.fa.art", "Arithmetic Compress","2.4 MB", "2.137s"],
        ["protein_seq.fa", "-","9.6 MB", "-"],
         ["protein_seq.fa.gz", "gzip","5.8 MB", "0.475s"], 
        ["protein_seq.fa.bz2", "bzip2","5.3 MB", "1.007s"],
        ["protein_seq.fa.bz2", "pbzip2","5.3 MB", "0.151s"],
        ["protein_seq.fa.art", "Arithmetic Compress","5.2 MB", "2.852s"]]
display(HTML(tabulate.tabulate(table, tablefmt='html')))

## Compressing real data

In [38]:
#>NC_023693.1:58516-60095 Enterobacteria phage phi92 gp120, complete genome
gp120_1 = "GGAGGTTCTGAGTGTCTACAGGTAAGCGTAAGTACACTAAAAGATCGGATTACTGGAACAAAGGCTCAAC
TGAAAAAGCTGCACTGTCTCCAACACAAAGTGCAACTAAAGAGAAGAACTTGGTACTTTCACCAGAAATA
GGTACTATTGGTCTTAATTCGATTAAAGCCTTTACGAACTTCATGCAACCGTATGAAACTCGCTTCCCAG
AAAACATCCGAACATACAAAGAAATGGGTGAAGACCCTGACGTAGCTACTGCTCTTGATGCTACGTACAT
CTTTGTAGATAGAGCATTCTTTGATTTTAAAATAAAATACAATGTTAGTTCAGCTAAGTCAAGACGTGCT
GCAAAGTTTGTAGACTACACACTTCGCAACATGAATGCACCACTAAGACAGTATGTACGCTCCTTGCTAA
CATATAAGCAATTTGGGTTTGCATTTGCAGAAAAGGTGTATGAACTTGATGAAGATCCAAAAAGCCCGTA
TTTTGGTTATTACCGTCTTGTTAAATTGGCGTTTAGACCCCAGGATACAATTGATTTGGCACAACCTTTC
ACTTATTCTGATGACGGTCGCACTATTCTTACGGTGAACCAAAACATTACTAATGGAATGGTAAGCCCTG
GTACTAACGCAACTCTAATAGGAAGAAAAGAAATTCCTATGGAGAAAGTTATTTACGTAGGGAGCAACAT
TACAGAGAATAATCCATTAGGTGTAAGCCCACTTCTTGCAGTATATAGAAGTTGGCGTGAAAAATCTCTG
ATTCAAGAATACGAAGTTGTTGGTGTTTCGAAAGACTTAGGCGGTATGCCAGTGTTGATGGTTCCAAGCG
ATATTCTCAACAGGGCATCACTAAACCCAAGCGGAGATGAAGCACAATCTTTACGTGTTCTGCAAGCAAA
TATCGCTAACTTACACGCTGGTGAACAATCTTATATGGTGTTACCTTCTGATGTTTATGAAGGTACTGTT
ATGCGCCAGTATGATCTTGTATTCCAAGGTGTTGAGGGTAGTGGTAAGCAATTCGATACACAGGCTCTAA
TAAAACAACGTAAATTAGATATCTACAACAGATTTGGTGCAGGTGTTCTGATTATGGGTGATGGAGAAGG
TGGTAGTTATTCTTTGTCAGATAACAAACAAACCCTTCTTTCACATTTTATTGAGCGTGATGTAGATATT
ATTACAGAAGCACTAAATACTCAAGTAATTCCCCAACTTCTACGTTTGAATGGTATTTTCTTATCTCAAG
AAGATATGCCTAAATTTGTTAGTGATGATATCGGTGATCCTGATATTGAAGTTAATGCTAAGGCAATCCA
ACAAATCGTAGCTGCTGGTGCAATACCATTAACGCCAGAAGTAATTAATGAATTCTTCGAACGTTTAGGT
TTCAATTATAGAATACCTGATGATATTGTTGCTGATCCAGATAAGTTCCAAGAATTCCTTGAGACTTTTA
TGCCAGATAAGACTTCAAGATCTGGTGATGGTTTAGCGGCTGGTGCTGGTAATGGAACGTCAACATCTCC
TGCTGCATTAGATACATCGGCAGCAAATTTGGCAAATTAA"

## Compress the multi-FASTAusing gzip, bzip2, and arithmetic coding

_A priori, doy ou expect to achieve better or worse compressionherethan random data? Why?_  
Worse because actual DNA sequences have more random sequences and are less structured and predictable. 


_How does the compression ratio of this file compare to random data?_  
    
    They all performed better compared to random data. 

In [13]:
print("Actual data: ")
Actual_data = [["File Name", "Compression method", "File size (KB)", "Compression ratio"],
        ["gp120_10_seq.fa", "-",2.7, 2.7/2.7],
               ["gp120_10_seq.fa.gz", "gzip",1.1, 2.7/1.1],
               ["gp120_10_seq.fa.bz2", "bzip2",.984, 2.7/.984],
               ["gp120_10_seq.fa.pz2", "pbzip2",.984, 2.7/.984],
               ["gp120_10_seq.fa.art", "Arithmetic Compress",2.1, 2.7/2.1]]


display(HTML(tabulate.tabulate(Actual_data, tablefmt='html')))

print("Random data: ")
Random_data = [["File Name", "Compression method", "File size (MB)", "Compression ratio"],
        ["nt_seq.fa", "-",9.6, 9.6/9.6],
         ["nt_seq.fa.gz", "gzip",2.8, 9.6/2.8], 
        ["nt_seq.fa.bz2", "bzip2",2.7, 9.6/2.7],
        ["nt_seq.fa.bz2", "pbzip2",2.7, 9.6/2.7],
        ["nt_seq.fa.art", "Arithmetic Compress",2.4, 9.6/2.4]]


display(HTML(tabulate.tabulate(Random_data, tablefmt='html')))
print("For every 2.7 bytes of data that you input, pbzip2 compresses that to 1.0 byte. Thus the fraction saved using pbzip2: ", .984/2.7)

Actual data: 


Random data: 


For every 2.7 bytes of data that you input, pbzip2 compresses that to 1.0 byte. Thus the fraction saved using pbzip2:  0.3644444444444444


## Estimating compression of 1000 terabytes

Let’s make some assumptions about the contents of the data at your biotech company. Most of the data, say 80%, isre-sequencing of genomes and plasmids that are very similar to each other. Another 10% might be protein sequences, and the last 10% are binary microscope images which we’ll assume follow the worst-case scenario of being completely random.

_Given the benchmarking data you obtained in this lab, which algorithm do you propose to use for each type of data? Provide an estimate for the fraction of space you can save using your compression scheme. How much of a bonus do you anticipate receiving this year?_

Use pbzip2 as it is the fastest. You can reduce the amount of storage to around 36.4% of its original space, approximately reducing storage by 2/3.

We will leave the 10% protein sequences and 10% are binary microscope images uncompressed since compression increases the file size.

Compression ratio for pbzip2: 2.72.
We use pbzip2 on all the data so  0.80 x .272 x 100 = 21.76%   
Reduction: 100% - 21.76% = 78.24%  
Amount savings: 50/terabyte * 0.7824 * 1000 terabytes/day * 365days/year = 14278800 dollars 
